### PR TITLE
feat(data): implement `SNBTDataFormat`

### PR DESCRIPTION
**Pull request by SettingDust**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/2432)

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/persistence/SNBTDataFormat.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+        return new BufferedReader(reader);
+    }
+
+    @Override
+    public DataContainer read(final String input) throws InvalidDataException, IOException {
+        try {
+            return NBTTranslator.INSTANCE.translate(TagParser.parseTag(input));
+        } catch (CommandSyntaxException e) {
+            throw new RuntimeException(e);
```
Comment (dualspiral) [inline src/main/java/org/spongepowered/common/data/persistence/SNBTDataFormat.java]: No - we should not be throwing `RuntimeException`s like this. Use `InvalidDataException` instead (same goes for all the following methods).

File: testplugins/src/main/java/org/spongepowered/test/data/DataFormatTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                                }}).key("format").build(),
+                                Parameter.remainingJoinedStrings().key("value").build())
+                        .executor(context -> {
+
+                            StringDataFormat format = context.requireOne(Parameter.key("format", StringDataFormat.class));
+                            String value = context.requireOne(Parameter.key("value", String.class));
+                            try {
+                                context.cause().audience().sendMessage(Component.text(format.write(format.read(value))));
+                            } catch (IOException e) {
+                                throw new RuntimeException(e);
```
Comment (dualspiral) [inline testplugins/src/main/java/org/spongepowered/test/data/DataFormatTest.java]: Don't throw `RuntimeExceptions` indiscriminately like this. In the case of commands, yes, it'll be caught, but the result is ugly. Gracefully handle it instead - we have `CommandException` that you should use instead if you must throw an exception.

File: src/main/java/org/spongepowered/common/data/persistence/SNBTDataFormat.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        return new BufferedReader(reader);
+    }
+
+    @Override
+    public DataContainer read(final String input) throws InvalidDataException, IOException {
+        try {
+            return NBTTranslator.INSTANCE.translate(TagParser.parseTag(input));
+        } catch (CommandSyntaxException e) {
```
Comment (dualspiral) [inline src/main/java/org/spongepowered/common/data/persistence/SNBTDataFormat.java]: `final` (same for all other catch blocks)

File: testplugins/src/main/java/org/spongepowered/test/data/DataFormatTest.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                        .addParameters(Parameter.choices(StringDataFormat.class, new HashMap<String, StringDataFormat>() {{
+                                    put("hocon", DataFormats.HOCON.get());
+                                    put("json", DataFormats.JSON.get());
+                                    put("snbt", DataFormats.SNBT.get());
+                                }}).key("format").build(),
+                                Parameter.remainingJoinedStrings().key("value").build())
+                        .executor(context -> {
+
+                            StringDataFormat format = context.requireOne(Parameter.key("format", StringDataFormat.class));
+                            String value = context.requireOne(Parameter.key("value", String.class));
```
Comment (dualspiral) [inline testplugins/src/main/java/org/spongepowered/test/data/DataFormatTest.java]: Declare as final (it's generally what we do in our codebase, so I'd like to see it here too)

File: src/main/java/org/spongepowered/common/data/persistence/SNBTDataFormat.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            return NBTTranslator.INSTANCE.translate(TagParser.parseTag(
+                    new BufferedReader(new InputStreamReader(input, StandardCharsets.UTF_8)).lines().collect(Collectors.joining("\n"))));
+        } catch (final CommandSyntaxException e) {
+            throw new InvalidDataException(e);
+        }
+    }
+
+    @Override
+    public String write(final DataView data) throws IOException {
+        return NBTTranslator.INSTANCE.translate(data).getPrettyDisplay().getString();
```
Comment (dualspiral) [inline src/main/java/org/spongepowered/common/data/persistence/SNBTDataFormat.java]: Why not `toString` instead of `getPrettyDisplay().getString()`?

Comment (SettingDust) [reply src/main/java/org/spongepowered/common/data/persistence/SNBTDataFormat.java]: `getPrettyDisplay().getString()` is "pretty". I noticed the hocon data format seems like is "pretty". So I change that. It's `toString` before

**Pull request merged**